flink-runtime/src/main/java/org/apache/flink/runtime/operators/hash/ReduceHashTable.java

```diff
@@ -0,0 +1,1048 @@
+	private static final int MIN_NUM_MEMORY_SEGMENTS = 3;
+
+	// Note: the following two constants can't be negative, because negative values are reserved for storing the
+	// negated size of the record, when it is abandoned (not part of any linked list).
+
+	/** The last link in the linked lists will have this as next pointer. */
+	private static final long END_OF_LIST = Long.MAX_VALUE;
+
+	/** This value means that prevElemPtr is "pointing to the bucket head", and not into the record segments. */
+	private static final long INVALID_PREV_POINTER = Long.MAX_VALUE - 1;
+
+
+	private static final long RECORD_OFFSET_IN_LINK = 8;
+
+
+	/** this is used by processRecordWithReduce */
+	private final ReduceFunction<T> reducer;
+
+	/** emit() sends data to outputCollector */
+	private final Collector<T> outputCollector;
+
+	private final boolean objectReuseEnabled;
+
+	/**
+	 * This initially contains all the memory we have, and then segments
+	 * are taken from it by bucketSegments, recordArea, and stagingSegments.
+	 */
+	private final ArrayList<MemorySegment> freeMemorySegments;
+
+	private final int numAllMemorySegments;
+
+	private final int segmentSize;
+
+	/**
+	 * These will contain the bucket heads.
+	 * The bucket heads are pointers to the linked lists containing the actual records.
+	 */
+	private MemorySegment[] bucketSegments;
+
+	private static final int bucketSize = 8, bucketSizeBits = 3;
+
+	private int numBuckets;
+	private int numBucketsMask;
+	private final int numBucketsPerSegment, numBucketsPerSegmentBits, numBucketsPerSegmentMask;
+
+	/**
+	 * The segments where the actual data is stored.
+	 */
+	private final RecordArea recordArea;
+
+	/**
+	 * Segments for the staging area.
+	 * (It should contain at most one record at all times.)
+	 */
+	private final ArrayList<MemorySegment> stagingSegments;
+	private final RandomAccessInputView stagingSegmentsInView;
+	private final StagingOutputView stagingSegmentsOutView;
+
+	private T reuse;
+
+	/** This is the internal prober that insertOrReplaceRecord and processRecordWithReduce use. */
+	private final HashTableProber<T> prober;
+
+	/** The number of elements currently held by the table. */
+	private long numElements = 0;
+
+	/** The number of bytes wasted by updates that couldn't overwrite the old record due to size change. */
+	private long holes = 0;
+
+	/**
+	 * If the serializer knows the size of the records, then we can calculate the optimal number of buckets
+	 * upfront, so we don't need resizes.
+	 */
+	private boolean enableResize;
+
+
+	/**
+	 * This constructor is for the case when will only call those operations that are also
+	 * present on CompactingHashTable.
+	 */
+	public ReduceHashTable(TypeSerializer<T> serializer, TypeComparator<T> comparator, List<MemorySegment> memory) {
+		this(serializer, comparator, memory, null, null, false);
+	}
+
+	public ReduceHashTable(TypeSerializer<T> serializer, TypeComparator<T> comparator, List<MemorySegment> memory,
```

Comment:
Can we remove the `ReduceFunction` and `Collector` from this class and move the corresponding logic into the driver?
It would be good if this table could also be used by a `CombineFunction` (not only a `ReduceFunction`). 
So, I would remove the following methods:
- `processRecordWithReduce()`: can be implemented by the driver using the Prober methods `getMatchFor()` and `updateMatch()`
- `emit()`: can be implemented by the driver using `getEntryIterator()`
- `emitAndReset()`: same as `emit()` but we need an additional `reset()` method

I would also rename the table if it becomes less specialized, maybe to `InPlaceMutableHashTable` or do you have a better idea, @ggevay?